mqtt/mqtt-broker/src/broker.rs

```diff
@@ -747,9 +789,59 @@ async fn publish_to(session: &mut Session, publication: &proto::Publication) ->
     Ok(())
 }
 
-impl Default for Broker {
-    fn default() -> Self {
-        Broker::new()
+#[derive(Clone, Debug, Default, PartialEq, Deserialize, Serialize)]
+pub struct BrokerState {
+    retained: HashMap<String, proto::Publication>,
+    sessions: Vec<SessionState>,
+}
+
+pub struct BrokerBuilder<N, Z> {
+    state: Option<BrokerState>,
+    authenticator: N,
+    authorizer: Z,
+}
+
+impl<N, Z> BrokerBuilder<N, Z>
+where
+    N: Authenticator,
+    Z: Authorizer,
+{
+    pub fn new(authenticator: N, authorizer: Z) -> Self {
+        Self {
+            state: Option::default(),
+            authenticator,
+            authorizer,
+        }
+    }
```

Comment:
I don't think the authenticator and authorizer need to be required here. We could default them to something that noops or denies everything be default. For example:

```rust
BrokerBuilder::new()
    .authenticator(authN)
    .authorizer(authZ)
    .state(state)
    .build()
```

```rust
impl<N, Z> BrokerBuilder<N, Z>
where
    N: Authenticator,
    Z: Authorizer,
{
    pub fn new() -> Self {
        Self {
            state: None,
            authenticator: DefaultAuthenticator,
            authorizer: DefaultAuthorizer,
        }
    }

    pub fn authenticator<N1: Authenticate>(mut self, authenticator: N1) -> BrokerBuilder<N1, Z> {
        BrokerBuilder {
            state: self.state,
            authenticator,
            authorizer: self.authorizer,
        }
    }

    ...
```